process/block/preprocess/smartContractResults.go

```diff
@@ -0,0 +1,424 @@
+	scr.scrForBlock.mutTxsForBlock.Lock()
+	scr.scrForBlock.txHashAndInfo[string(smartContractResultHash)] = &txInfo{tx: smartContractResult, txShardInfo: txShardInfo, has: true}
+	scr.scrForBlock.mutTxsForBlock.Unlock()
+
+	return nil
+}
+
+// RequestSmartContractResultsForMiniBlock requests missing smartContractResults for a certain miniblock
+func (scr *smartContractResults) RequestTransactionsForMiniBlock(mb block.MiniBlock) int {
+	missingScrsForMiniBlock := scr.computeMissingScrsForMiniBlock(mb)
+	scr.onRequestSmartContractResult(mb.SenderShardID, missingScrsForMiniBlock)
+
+	return len(missingScrsForMiniBlock)
+}
+
+// computeMissingScrsForMiniBlock computes missing smartContractResults for a certain miniblock
+func (scr *smartContractResults) computeMissingScrsForMiniBlock(mb block.MiniBlock) [][]byte {
+	missingSmartContractResults := make([][]byte, 0)
+	if mb.Type != block.SmartContractResultBlock {
+		return missingSmartContractResults
+	}
+
+	for _, txHash := range mb.TxHashes {
+		tx := scr.getTransactionFromPool(mb.SenderShardID, mb.ReceiverShardID, txHash, scr.scrPool)
+		if tx == nil {
+			missingSmartContractResults = append(missingSmartContractResults, txHash)
+		}
+	}
+
+	return missingSmartContractResults
+}
+
+// getAllScrsFromMiniBlock gets all the smartContractResults from a miniblock into a new structure
+func (scr *smartContractResults) getAllScrsFromMiniBlock(
+	mb *block.MiniBlock,
+	haveTime func() bool,
+) ([]*smartContractResult.SmartContractResult, [][]byte, error) {
+
+	strCache := process.ShardCacherIdentifier(mb.SenderShardID, mb.ReceiverShardID)
+	txCache := scr.scrPool.ShardDataStore(strCache)
+	if txCache == nil {
+		return nil, nil, process.ErrNilScrDataPool
+	}
+
+	// verify if all smartContractResult exists
+	smartContractResults := make([]*smartContractResult.SmartContractResult, 0)
+	txHashes := make([][]byte, 0)
+	for _, txHash := range mb.TxHashes {
+		if !haveTime() {
+			return nil, nil, process.ErrTimeIsOut
+		}
+
+		tmp, _ := txCache.Peek(txHash)
+		if tmp == nil {
+			return nil, nil, process.ErrNilSmartContractResult
+		}
+
+		tx, ok := tmp.(*smartContractResult.SmartContractResult)
+		if !ok {
+			return nil, nil, process.ErrWrongTypeAssertion
+		}
+
+		txHashes = append(txHashes, txHash)
+		smartContractResults = append(smartContractResults, tx)
+	}
+
+	return smartContractResults, txHashes, nil
+}
+
+// CreateAndProcessMiniBlock creates the miniblock from storage and processes the smartContractResults added into the miniblock
+func (scr *smartContractResults) CreateAndProcessMiniBlock(sndShardId, dstShardId uint32, spaceRemained int, haveTime func() bool, round uint32) (*block.MiniBlock, error) {
+	return nil, nil
+}
+
+// ProcessMiniBlock processes all the smartContractResults from a and saves the processed smartContractResults in local cache complete miniblock
+func (scr *smartContractResults) ProcessMiniBlock(miniBlock *block.MiniBlock, haveTime func() bool, round uint32) error {
```

Comment:
This method is not tested at all

Reply:
were copied from blockprocessor, not tested there with unit tests. Unit tests will be added. Added TODO

Reply:
ok